README.adoc

```diff
@@ -82,6 +82,64 @@ Once that is finished, the wallet registry can call
 `Wallets.registerNewWallet`, which sets the new wallet as the active wallet.
 Going forward, it will receive deposits.
 
+==== Wallet Closure
+
+Wallets can close in a few ways:
+
+* The operators notify the chain that the wallet is failing a heartbeat: `Wallets.notifyWalletHeartbeatFailed`
+* Someone notifies the chain that the wallet timed out while filling a redemption: `Wallets.notifyWalletTimedOutRedemption`
+* Someone notifies the chain that the wallet is too old: `Wallets.notifyCloseableWallet`
+* Someone notifies the chain that the wallet has too few BTC remaining: `Wallets.notifyCloseableWallet`
+
+Each of the above routes into `Wallets.moveFunds`. If the wallet has no main
+UTXO, then we can start closing it immediately via
+`Wallets.beginWalletClosing`. Otherwise, we change it's state to
+`WalletState.MovingFunds`, decrease the live wallet count, and give the wallet
+a deadline to move its funds to other wallet(s). If this was the active wallet,
+then we currently have no active wallet.
```

Comment:
The active wallet can not be closed by `notifyCloseableWallet`:
https://github.com/keep-network/tbtc-v2/blob/f1076f450a05c6b87c91e2f49a51fb43799f5439/solidity/contracts/bridge/Wallets.sol#L343-L346

We require another wallet to be created to replace it.

Reply:
well hot dang!

Reply:
The other routes don't have that protection though:

https://github.com/keep-network/tbtc-v2/blob/main/solidity/contracts/bridge/Wallets.sol#L399

If we get into `movingFunds` via `Bridge.notifyRedemptionTimeout` or `WalletRegistry.notifyOperatorInactivity` and the wallet is the active wallet, then we're left with no active wallet. I'll amend the last two bullet points!

Reply:
[b08e80a](https://github.com/keep-network/tbtc-v2/pull/327/commits/b08e80a38d1f2dd31e897fd9f963f7e1902fd270)

Reply:
Yeah, it's only for `notifyCloseableWallet`. And I think it makes sense - if the wallet is in a good shape (no heartbeat failure), we can still rely on it. Rotating active wallets just because they are old enough and without giving them a chance to accumulate enough wealth could lead to too many move funds operations between wallets.